                                       CAPSTONE PROJECT
                                              ON
                                 ANALYZE THE CUSTOMER REVIEWS
                                              BY   
                                          DEVIKA ENUGU

# 1 PROBLEM STATEMENT –

▪ You are working in an e-commerce company, and your company has put forward a task to analyze the customer reviews for various products. You are supposed to create a report that classifies the products based on the customer reviews.

▪ Dataset Information: The Reviews.csv dataset contains 568454 rows and 10 columns.

# 2 PROJECT OBJECTIVE –



  ▪ Product reviews are a major reason why consumers prefer ecommerce. Where else can you go to read hundreds of candid testimonials or complaints about every aspect of an item.

2.1 Consumers Prefer Online Review –

 ▪ Study after study has shown that consumers are increasingly reliant upon online reviews when making purchasing decisions.

 ▪ In fact, a recent Nielson study found that 70 percent of global consumers trust online reviews, up 15 percent within a four-year period. Industry experts project this trend to further increase in the coming years.

 ▪ If consumers love online reviews, what better reason does a merchant need? Sellers are in the business of delivering product that customers will rave about.
 
 ▪ Acknowledging the growing importance of online reviews is the first step toward better understanding the market.

2.2 Reviews Provide Valuable Market Intelligence-

 ▪ For Entrepreneurs new to ecommerce, determining which product to sell can be overwhelming. This decision is even difficult for the most experienced sellers. To bring order to this process, many merchants turn to reviewing and tracking reviews.
    
 ▪ Consider, for a moment, Amazon’s full directory of product categories, within a given category, there are several subcategories. At the product level, each item is given “Amazon Best Sellers Rank’ within Subcategories. By analyzing the reviews from product at the top, middle and bottom of these lists, sellers can gain invaluable insights into consumers preference and concerns.

 ▪ In short, reviews can serve as a digital focus group for any given product. This freely available information can then be used by the merchant to make inventor or strategic placement decision and even to create entirely new product and brands to address limitation in current offering.

2.3 Smart Sellers Monitor Reviews-

 ▪ Reviews play even more important role for third-party merchants.

 ▪ Understanding the psychology of customer reviews is important for making smart business decision, winning the Buy Box and ultimately thriving online.

2.4 Customers Can Vent About Product, Not the Merchant-

 ▪ Although product can certainly be a representation of the merchant’s brand, it is impossible for a seller to ensure satisfaction for each item sold.

 ▪ Personal taste and preference, incorrect sizing or unrealistic expectation can dissonance.

 ▪ A negative comment about the product in the form of a review is typically less impactful to the seller than negative feedback.

 ▪ By offering two separate forums for people to share their opinions, merchants are able to protect their reputation.

# 3 DATA DESCRIPTION –

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('Reviews_2.csv')
df.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...  
3         Cough Medicine  If you are looking for the secret ingredient i...  
4            Great taffy  Great taffy at a great price.  There was a wid...

▪ There is 10 attribute Id, ProductId, UserId, ProfileName, HelpfullnessNumerator, HelpfullnessDenominator, Score, Time, Summary and Text.

▪ For further information about the data, we can get the column name, the shape of the data and the dimensions of data available to have more clarity.

In [3]:
df.shape

(568454, 10)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 568454 entries, 0 to 568453
Data columns (total 10 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Id                      568454 non-null  int64 
 1   ProductId               568454 non-null  object
 2   UserId                  568454 non-null  object
 3   ProfileName             568438 non-null  object
 4   HelpfulnessNumerator    568454 non-null  int64 
 5   HelpfulnessDenominator  568454 non-null  int64 
 6   Score                   568454 non-null  int64 
 7   Time                    568454 non-null  int64 
 8   Summary                 568427 non-null  object
 9   Text                    568454 non-null  object
dtypes: int64(5), object(5)
memory usage: 43.4+ MB


▪ The info () method is useful to get a quick description of the data, in particular the total
number of rows, and each attribute’s type and number of non- null values

▪ There are 568454 instances in the dataset. Notice that the ProfileName, Summary attribute has 568438 ,5688427 non null values, meaning that 16 ProfileName and 27 Summary missing this feature.

▪ Let’s look at the other fields. The describe () method shows a summary of the numeric attributes.

In [5]:
df.describe()

Id  HelpfulnessNumerator  HelpfulnessDenominator  \
count  568454.000000         568454.000000            568454.00000   
mean   284227.500000              1.743817                 2.22881   
std    164098.679298              7.636513                 8.28974   
min         1.000000              0.000000                 0.00000   
25%    142114.250000              0.000000                 0.00000   
50%    284227.500000              0.000000                 1.00000   
75%    426340.750000              2.000000                 2.00000   
max    568454.000000            866.000000               923.00000   

               Score          Time  
count  568454.000000  5.684540e+05  
mean        4.183199  1.296257e+09  
std         1.310436  4.804331e+07  
min         1.000000  9.393408e+08  
25%         4.000000  1.271290e+09  
50%         5.000000  1.311120e+09  
75%         5.000000  1.332720e+09  
max         5.000000  1.351210e+09

▪ The count, mean, min and max rows are self-explanatory. Note that null values are ignored.

▪ The std rows shows the standard deviation, which measure how dispersed the values are.

▪ The 25 %, 50 % and 75 % rows show the corresponding percentiles; percentiles indicate the value below which a given percentage of observation in a group of observation falls.

# 4 DATA PRE-PROCESSING STEPS AND INSPIRATION –

4.1 Loading The Data –
 
    We have imported all the necessary libraries to perform the EDA, and have imported the dataset using the read_csv () from pandas module. EDA is an approach to analyze the data using visual techniques. It is used to discover trends, patterns, or to check assumptions with the help of statistical summary and graphical representations. 

4.2 Data Cleaning and Encoding– 

   • Cleaning up the data frame, by dropping any rows that have missing values.
    
   • Most Machine Learning algorithm cannot work with missing feature, so that created a function to take care of them. We noticed earlier that There are 568454 instances in the dataset, ProfileName, Summary attribute has 568438 ,5688427 non null values, meaning that 16 ProfileName and 27 Summary missing this feature.

   • We can accomplish these easily using Data Frame’s dropna () methods.
  
   • The Score column is scaled from 1 to 5, and we will remove all Scores equal to 3 because we assume these are neutral and did not provide us any useful information. We then add a new column called “Positivity”, where any score above 3 is encoded as a 1, indicating it was positively rated. Otherwise, it’ll be encoded as a 0, indicating it was negatively rated.


In [6]:
df.isnull().sum()

Id                         0
ProductId                  0
UserId                     0
ProfileName               16
HelpfulnessNumerator       0
HelpfulnessDenominator     0
Score                      0
Time                       0
Summary                   27
Text                       0
dtype: int64

In [7]:
df.dropna(inplace=True)

In [8]:
df.isnull().sum()

Id                        0
ProductId                 0
UserId                    0
ProfileName               0
HelpfulnessNumerator      0
HelpfulnessDenominator    0
Score                     0
Time                      0
Summary                   0
Text                      0
dtype: int64

In [9]:
df.shape

(568411, 10)

In [10]:
df.columns

Index(['Id', 'ProductId', 'UserId', 'ProfileName', 'HelpfulnessNumerator',
       'HelpfulnessDenominator', 'Score', 'Time', 'Summary', 'Text'],
      dtype='object')

In [11]:
df.dtypes

Id                         int64
ProductId                 object
UserId                    object
ProfileName               object
HelpfulnessNumerator       int64
HelpfulnessDenominator     int64
Score                      int64
Time                       int64
Summary                   object
Text                      object
dtype: object

In [12]:
df[df['Score'] != 3]
df['Positivity'] = np.where(df['Score']>3,1,0)
df.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   

                 Summary                                               Text  \
0  Good Quality Dog Food  I have bought several of the Vitality canned d...   
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...   
2  "Delight" says it all  This is a confection that has been around a fe...   
3         Cough Medicine  If you are looking for the secret ingredient i...   
4            Great taffy  Great taffy at a great price.  There was a wid...   

   Positivity  
0           1  
1           0  
2           1  
3           0  
4           1

# 4.3 Splitting The Data into Train and Test –

▪ Scikit-Learn provide a few functions to split dataset into multiple subsets in various ways. The simplest function is train_test_split, which doesn’t pretty much same thing as the function split_train_test, with a couple of additional features.

▪ First there is random_state parameter that allows to set the random generator seed.

▪ Second, we can pass it multiple datasets with an identical number of rows, and it will split them on the same indices. 

▪ Split our data into random training and test subsets using “Text” and “Positivity” columns, and then print out the first entry and the shape of the training set. 

▪ Looking at X_train, we can see that we have a collection of over 426308 reviews or documents.

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
x_train,x_test,y_train,y_test = train_test_split(df['Text'],df['Positivity'],random_state=0)

In [15]:
print('x_train first entry: \n\n',x_train[0])

x_train first entry: 

 I have bought several of the Vitality canned dog food products and have found them all to be of good quality. The product looks more like a stew than a processed meat and it smells better. My Labrador is finicky and she appreciates this product better than  most.


In [16]:
x_train.shape

(426308,)

In [17]:
x_test.shape

(142103,)

In [18]:
y_train.shape

(426308,)

In [19]:
y_test.shape

(142103,)

▪ In order to perform machine learning on text documents, we first need to turn these text content into numerical feature vectors that Scikit-Learn can use.

4.4 Bag-of-Words Model – 

▪ The simplest and most intuitive way to do so, is the “bags-of-words” representation which ignores structure and simply counts how often each word occurs.

▪ CountVectorizer allows us to use the bags-of-words approach, by converting a collection of text documents into a matrix of token counts. 

▪ We instantiate the CountVectorizer and fit it to our training data, converting our collection of text documents into a matrix of token counts.

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
vect = CountVectorizer().fit(x_train)
vect

CountVectorizer()

4.5 Default Ton figuration Tokenize-

▪ This model has many parameters; however, the default values are quite reasonable for our purpose. 

▪ The default configuration tokenizes the string, by extracting words of at least 2 letters or numbers, separated by word boundaries, it then converts everything to lowercase and builds a vocabulary using these tokens.

▪ We can get some of the vocabularies by using the get_feature_names method.

In [21]:
vect.get_feature_names()[::2000]

C:\Users\hp\anaconda3\Library\mkspecs\devices\devika\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['00',
 '255g',
 '843mg',
 'aftertraste',
 'anticarcinogens',
 'average',
 'b000mrd5jo',
 'b001rqemwi',
 'b005jd60wk',
 'beleive',
 'boobs',
 'buttersworth',
 'cc',
 'chuy',
 'compresses',
 'cramper',
 'decap',
 'difficulkt',
 'dreamy',
 'enchanted',
 'expedited',
 'fists',
 'frother',
 'gloved',
 'gurantees',
 'hiking_',
 'images',
 'intruder',
 'kavanagh',
 'lawry',
 'lowry',
 'matured',
 'misnomer',
 'mythreads',
 'numorous',
 'osco',
 'paupua',
 'pittston',
 'preshave',
 'quart',
 'refrigerante',
 'ringworm',
 'savedge',
 'sheer',
 'smiths',
 'sprklng',
 'subtotal',
 'taos',
 'tiis',
 'tubed',
 'unsuccessful',
 'vomitar',
 'wintery',
 'zest']

In [22]:
len(vect.get_feature_names())

106260

4.6 Transform The Documents In x_train To a Document Term Matrix -

▪ Transform gives us the bags-of-word representation of X_train.

▪ The result is stored in a SciPy sparse matrix, where each row corresponds to a document, and each column is a word from our training vocabulary

In [23]:
x_train_vectorized = vect.transform(x_train)
x_train_vectorized

<426308x106260 sparse matrix of type '<class 'numpy.int64'>'
	with 22990341 stored elements in Compressed Sparse Row format>

4.7 Train The Logistic Regression Classifier-

▪ Train the Logistic Regression classifier based on this feature matrix X_ train_ vectorized, because Logistics Regression works well for high dimensional sparse data.

In [24]:
import warnings
warnings.filterwarnings('ignore')

In [25]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(x_train_vectorized, y_train)

LogisticRegression()

4.8 Predictions Using X_test-

▪ Make predictions using X_test, and compute the area under the curve score.

▪ determine coefficients weights-

In [26]:
predictions = model.predict(vect.transform(x_test))

In [27]:
predictions

array([1, 1, 1, ..., 0, 1, 0])

In [28]:
from sklearn.metrics import roc_auc_score
print('AUC: ', roc_auc_score(y_test,predictions))

AUC:  0.8119806401630989


4.9 Coefficients For Each Feature 

▪ In order to better understand how our model makes these predictions, we can use the coefficients for each feature (a word) to determine its weight in terms of positivity and negativity.

In [29]:
feature_names = np.array(vect.get_feature_names())
sorted_coef_index = model.coef_[0].argsort()
print('Smallest Coefs: \n{}\n'.format(feature_names[sorted_coef_index[:10]]))
print('Largest Coefs: \n{}\n'.format(feature_names[sorted_coef_index[:-11:-1]]))

Smallest Coefs: 
['worst' 'disappointing' 'disappointment' 'awful' 'terrible' 'disgusting'
 'threw' 'yuck' 'sorry' 'tasteless']

Largest Coefs: 
['hooked' 'pleasantly' 'pleased' 'skeptical' 'yum' 'beat' 'awesome'
 'satisfied' 'yummy' 'delicious']



4.10 Tf–idf Term Weighting-

▪ In a large text corpus, some words will be present very often but will carry very little meaningful information about the actual contents of the document (such as “the”,
“a” and “is”). If we were to feed the count data directly to a classifier those very frequent terms would shadow the frequencies of rarer yet more interesting terms. 

▪ Tf-idf allows us to weight terms based on how important they are to a document.

▪ Specify min_df = 5, which will remove any words from our vocabulary that appear in fewer than five documents.

In [30]:
from sklearn.feature_extraction.text import TfidfVectorizer
vect = TfidfVectorizer(min_df = 5).fit(x_train)
len(vect.get_feature_names())

36692

In [31]:
x_train_vectorized = vect.transform(x_train)
model = LogisticRegression()
model.fit(x_train_vectorized, y_train)
predictions = model.predict(vect.transform(x_test))
print('AUC: ', roc_auc_score(y_test,predictions))

AUC:  0.8294548648029373


4.11 Obtain Features with Tf-Idf-

▪ Obtain a list of features with the smallest tf-idf that either commonly appeared across all reviews or only appeared rarely in very long reviews and a list of features with the largest tf–idf contains words which appeared frequently in a review, but did not appear commonly across all reviews.

In [32]:
feature_names = np.array(vect.get_feature_names())
sorted_tfidf_index = x_train_vectorized.max(0).toarray()[0].argsort()
print('Smallest Tfidf: \n{}\n'.format(feature_names[sorted_tfidf_index[:10]]))
print('Largest Tfidf: \n{}\n'.format(feature_names[sorted_tfidf_index[:-11:-1]]))

Smallest Tfidf: 
['0373' 'amortized' '0004' '1534' '231' 'amortization' '0377'
 'furnishings' '368' 'serine']

Largest Tfidf: 
['carmel' '98' 'yum' 'good' 'filler' 'word' 'love' 'mmm' 'awesome'
 'banana']



# 5. CHOOSING THE ALGORITHM FOR THE PROJECT

▪ Logistics Regression works well for high dimensional sparse data. 

▪ Logistic Regression is commonly used to estimate the probability that an instance belongs to a particular class. If the estimated probability is greater than 50 % then the model predicts that the instance belongs to that class (called the positive class, labeled “1”), or else it predicts that it does not (i.e., it belongs to the negative class, labeled “0”). This makes it binary classifier. 

▪ In the Score column is scaled from 1 to 5, and we will remove all Scores equal to 3 because we assume these are neutral and did not provide us any useful information. 

▪ We then add a new column called “Positivity”, where any score above 3 is encoded as a 1, indicating it was positively rated. Otherwise, it’ll be encoded as a 0, indicating it was negatively rated.

# 6. ASSUMPTIONS

▪ A linear relationship between the dependent and independent variables Not require 

▪ The error terms (residuals) do not need to be normally distributed. 

▪ Homoscedasticity is not required. 

▪ The dependent variable is not measured on an interval or ratio scale. 

▪ Observations to be independent of each other. In other words, the observations should not come from repeated measurements or matched data. 

# 7.MODEL EVALUATION AND TECHNIQUES

.Import all the necessary libraries

.Import the dataset

.Dropping any rows that have missing values

.Encode The Score column is

.Add a new column called “Positivity”

.Splitting The Data into Train and Test

.Instantiate the CountVectorizer which allows us to use the bags-of-words approach, by converting a collection of text documents into a matrix of token counts.

.Use get_feature_names method to get the vocabularies.

.Transform the documents in X_train to a document term matrix

.Train the Logistic Regression classifier

.Use the coefficients for each feature (a word) to determine its weight in terms of positivity and negativity

.Instantiate the tf–idf vectorizer to weight terms-based importance

.Obtain a list of features with the smallest tf-idf and largest tf–idf

# 8. INFERENCES FROM THE PROJECT

. There may be a correlation between the length of a review and the star rating it gives. Longer reviews may tend to have more detailed feedback and possibly higher star ratings.

. Certain products may have more negative or positive reviews than others, indicating that there are factors that influence customer satisfaction beyond just the quality of the product itself, such as shipping speed, customer service, or packaging.

. Some reviews may be fake or incentivized, either from competitors or from companies themselves trying to artificially inflate their product ratings. Detecting these fake reviews could be important for maintaining the integrity of the review system on Amazon.

. There may be differences in review patterns or sentiment based on geographic location, cultural background, or other demographic factors of the reviewers.

# 9. FUTURE POSSIBILITIES OF THE PROJECT 

▪ More and more people will be use artificial intelligence tools for ecommerce to boost conversions. 

▪ Customers will believe their experiences could be improved if the pages and links weren’t so small to click on. Another concern for these customers is the lack of information and product reviews available on some online platforms.

▪ These areas of improvement will make mobile ecommerce one of the hottest ecommerce future trends that companies could really work on. 

▪ The introduction of AI and machine learning will here to revolutionize the future of ecommerce. Ecommerce stores will be incorporating machine learning technology on their websites to enhance their customers’ experiences.

▪ Machine learning will be starting to play a major role in the ecommerce industry. With this technology businesses will show product recommendations based on customer’s buying behavior and the pages they visit. Moreover, AI and machine learning will help to give customers a visual experience of the product in the form of 3D models, online tours, etc. 

▪ When it comes to the future of e-business, it’s safe to say that the use of machine learning to enhance online customer experiences will become one of the hottest ecommerce trends, one that ecommerce businesses will greatly benefit from.

# CONCLUSION

The analysis reveals two group of reviewers standing out in helpfulness to the community: First group are those who write long reviews, and second are the ones writing very frequent shorter reviews. For a simple logistic regression electronic review dataset can accuracy close to 0.8 was possible

# 10. REFERENCES

kaggle website